obspy/io/nordic/ellipse.py

```diff
@@ -0,0 +1,329 @@
+Still missing:
+ - ellipsoids (3D ellipses)
+ - to_uncert()  (convert ellipse to covariance matrice, needed to
+                 convert QuakeML to nordic, for example)
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+import copy
+
+eps = np.finfo(float).eps
+
+
+class _ellipse:
+    def __init__(self, a, b, theta=0, center=(0, 0)):
+        """Defines an ellipse
+
+        The ellipse is assumed to be centered at zero with its semi-major axis
+        axis aligned along the NORTH axis (geographic standard, not math
+        standard!) unless orientation and/or center are set otherwise
+        You can think of it as theta (geographic angle) = 90-phi (math angle)
+
+        Inputs:
+            a=length of semi-major axis
+            b=length of semi-minor axis
+            theta=azimuth (degrees CW from 0=N (y positive))
+            center=x,y coordinates of ellipse center (0,0)
+        """
+        if a < b:
+            print('Semi-major smaller than semi-minor! Switching...')
+            self.a = b
+            self.b = a
+        else:
+            self.a = a
+            self.b = b
+        self.theta = theta
+        self.x = center[0]
+        self.y = center[1]
+
+    @classmethod
+    def from_cov(cls, cov, center=(0, 0), debug=False):
+        """Set error ellipse using covariance matrix
+
+        Call as e=_ellipse.from_cov(cov,center)
+
+        Inputs:
+            cov: covariance matrix [[c_xx, c_xy], [c_xy, c_yy]]
+            center:  center position (x,y)
+        Sources:
+            http://www.visiondummy.com/2014/04/
+                   draw-error-ellipse-representing-covariance-matrix/
+            https://blogs.sas.com/content/iml/2014/07/23/
+                    prediction-ellipses-from-covariance.html
+        """
+        evals, evecs = np.linalg.eig(cov)
+        evals = np.sqrt(evals)
+        if debug:
+            print(evecs)
+            print(evals)
+        # Sort eigenvalues in decreasing order
+        sort_indices = np.argsort(evals)[::-1]
+        # Select semi-major and semi-minor axes
+        a, b = evals[sort_indices[0]], evals[sort_indices[1]]
+        # Calculate angle of semi-major axis
+        x_v1, y_v1 = evecs[:, sort_indices[0]]
+        if y_v1 == 0.:
+            theta = 90.
+        else:
+            theta = (np.degrees(np.arctan((x_v1)/(y_v1))) + 180) % 180
+        return cls(a, b, theta, center)
+
+    @classmethod
+    def from_uncerts(cls, x_err, y_err, c_xy, center=(0, 0), debug=False):
+        """Set error ellipse using Nordic epicenter uncertainties
+
+        Call as e=_ellipse.from_uncerts(x_err,y_err,c_xy,center)
+
+        Inputs:
+            x_err: x error (m)
+            y_err: y error (m)
+            c_xy:  x-y covariance (m^2)
+            center:  center position (x,y)
+        """
+        cov = [[x_err**2, c_xy], [c_xy, y_err**2]]
+        return cls.from_cov(cov, center)
+
+    @classmethod
+    def from_uncerts_baz(cls, x_err, y_err, c_xy, dist, baz,
```

Comment:
It looks like this isn't covered by a test.

Reply:
``test_ellipse_from_uncerts_baz()`` added